### PR TITLE
refactor(web): use queries and TypeScript in storage components

### PR DESCRIPTION
**Pull request by imobachgs**
Merge the following pull requests into `master`:

* #1577 
* #1578 
* #1580

### CODE REVIEW

File: web/src/components/storage/ProposalActionsDialog.tsx [hunk 1]
```diff
@@ -71,6 +78,7 @@ export default function ProposalActionsDialog({ actions = [] }) {
         subvolActions.length,
       );
 
+  console.log(actions);
```
Comment (dgdavid) [inline web/src/components/storage/ProposalActionsDialog.tsx:81]: Debugging left over, I guess

File: web/src/components/storage/ProposalPage.test.tsx [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+ *
+ * To contact SUSE LLC about this file by physical or electronic mail, you may
+ * find current contact information at www.suse.com.
+ */
+
+/*
+ * NOTE: this test is not useful. The ProposalPage loads several queries but,
+ * perhaps, each nested component should be responsible for loading the
+ * information they need.
+ */
```
Comment (dgdavid) [inline web/src/components/storage/ProposalPage.test.tsx:26]: :+1:

File: web/src/components/storage/ProposalPage.tsx [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+// value: list of items which can be changed without affecting
+//        the state of the component
+export const NOT_AFFECTED = {
+  // the EncryptionField shows the skeleton only during initial load,
+  // it does not depend on any changed item and does not show skeleton later.
+  // the ProposalResultSection is refreshed always
+  InstallationDeviceField: [CHANGING.ENCRYPTION, CHANGING.BOOT, CHANGING.POLICY, CHANGING.VOLUMES],
+  PartitionsField: [CHANGING.ENCRYPTION, CHANGING.POLICY],
+  ProposalActionsSummary: [CHANGING.ENCRYPTION, CHANGING.TARGET],
+};
```
Comment (dgdavid) [inline web/src/components/storage/ProposalPage.tsx:70]: That reminds me I'd like to look for another approach for this now that the UI has changed A LOT.

File: web/src/components/storage/SpacePolicySelection.tsx [hunk 4]
```diff
@@ -131,10 +119,10 @@ export default function SpacePolicySelection() {
 
   // Resets actions (i.e., sets everything to "keep") if the custom policy has not been used yet.
   useEffect(() => {
-    if (policy?.id !== "custom" && !customUsed) setActions([]);
+    if (policy && policy?.id !== "custom" && !customUsed) setActions([]);
```
Comment (dgdavid) [inline web/src/components/storage/SpacePolicySelection.tsx:122]: Nice catch!

File: web/src/components/storage/SpacePolicySelection.tsx [hunk 5]
```diff
@@ -84,37 +82,27 @@ const SpacePolicyPicker = ({ currentPolicy, onChange = noop }) => {
  * Renders a page that allows the user to select the space policy and actions.
  */
 export default function SpacePolicySelection() {
-  const [state, setState] = useState({ load: false, settings: {} });
-  /** @type ReturnType<typeof useState<SpacePolicy>> */
-  const [policy, setPolicy] = useState();
+  const { settings } = useProposalResult();
+  const updateProposal = useProposalMutation();
```
Comment (dgdavid) [inline web/src/components/storage/SpacePolicySelection.tsx:86]: NP: just a note from my side for this and similar _mutation hooks_. A matter of taste, but I use to write them in a different way:

```suggestion
  const { mutateAsync: updateProposal } = useProposalMutation();
```

And then just use the `updateProposal(withWhatever)`.

But as said, just a **personal** preference.

File: web/src/queries/storage.ts [hunk 6]
```diff
@@ -0,0 +1,349 @@
...
+  const devices = useDevices("system", { suspense: true });
+  const { data } = useSuspenseQuery(usableDevicesQuery);
+
+  return data.map((sid) => findDevice(devices, sid)).filter((d) => d);
+};
+
+/**
+ * Hook that returns the product parameters (e.g., mount points).
+ */
+const useProductParams = (options?: QueryHookOptions): ProductParams => {
```
Comment (dgdavid) [inline web/src/queries/storage.ts:118]: Interesting... should it be here or in the software queries?

File: web/src/queries/storage.ts [hunk 7]
```diff
@@ -0,0 +1,349 @@
...
+  return results.map(({ data }) => buildVolume(data, systemDevices, product.mountPoints));
+};
+
+/**
+ * Hook that returns the devices that can be selected as target for volume.
+ *
+ * A device can be selected as target for a volume if either it is an available device for
+ * installation or it is a device built over the available devices for installation. For example,
+ * a MD RAID is a possible target only if all its members are available devices or children of the
+ * available devices.
```
Comment (dgdavid) [inline web/src/queries/storage.ts:144]: Me reading storage stuff: :exploding_head:

File: web/src/queries/storage.ts [hunk 8]
```diff
@@ -0,0 +1,349 @@
...
+  return {
+    settings: {
+      ...settings,
+      targetPVDevices: settings.targetPVDevices || [],
+      target: buildTarget(settings.target),
+      volumes: settings.volumes.map((v) => buildVolume(v, systemDevices, productMountPoints)),
+      // NOTE: strictly speaking, installation devices does not belong to the settings. It
+      // should be a separate method instead of an attribute in the settings object.
+      // Nevertheless, it was added here for simplicity and to avoid passing more props in some
+      // react components. Please, do not use settings as a jumble.
```
Comment (dgdavid) [inline web/src/queries/storage.ts:233]: :100:

File: web/src/types/storage.ts [hunk 9]
```diff
@@ -0,0 +1,194 @@
+/*
```
Comment (dgdavid) [inline web/src/types/storage.ts:1]: Should we explain the differences between `api/storage/types` and `types/storage` somewhere? What is more, should we have a plan for _unification_?

File: web/src/utils.js [hunk 10]
```diff
@@ -298,18 +295,16 @@ const hex = (value) => {
 
 /**
  * Converts an issue to a validation error
- * @function
  *
  * @todo This conversion will not be needed after adapting Section to directly work with issues.
  *
- * @param {import("~/client/mixins").Issue} issue
+ * @param {import("~/types/issues").Issue} issue
```
Comment (dgdavid) [inline web/src/utils.js:301]: I wonder if would be better to import types at the top of the file.

**Review by dgdavid — APPROVED** [review]
LGTM, just a few questions and notes plus a `console.log` leftover there.

I'll approve, feel free to merge it without needing another review if you don't introduce big changes.

Thank you for addressing this one!

File: web/src/components/storage/BootSelection.tsx [hunk 11]
```diff
@@ -49,78 +44,61 @@ const BOOT_DISABLED_ID = "boot-disabled";
...
+      selectedOption = BOOT_AUTO_ID;
+    } else {
+      selectedOption = BOOT_MANUAL_ID;
+    }
+
+    const findDevice = (name: string) => availableDevices.find((d) => d.name === name);
 
-    load().catch(console.error);
-  }, [state, loadAvailableDevices, loadProposalResult]);
+    setState({
```
[the remark this answers is not in the document]
Comment (imobachgs) [reply web/src/components/storage/BootSelection.tsx:78]: Yes, please, let's use another PR. Thanks!

File: web/src/api/storage/proposal.ts [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, contact SUSE LLC.
+ *
+ * To contact SUSE LLC about this file by physical or electronic mail, you may
+ * find current contact information at www.suse.com.
+ */
+
+import { get, put } from "../http";
+import { Action, ProductParams, ProposalSettings, ProposalSettingsPatch, Volume } from "./types";
```
[the remark this answers is not in the document]
Comment (imobachgs) [reply web/src/api/storage/proposal.ts:23]: Well, in this case it should be `~/api/storage/types`, but you are right.

File: web/src/api/storage/types.ts [hunk 13]
```diff
@@ -0,0 +1,416 @@
+// This file is auto-generated by @hey-api/openapi-ts
```
[the remark this answers is not in the document]
Comment (imobachgs) [reply web/src/api/storage/types.ts:1]: I will add the copyright. We might need to think about that when we automate the generation of such a file.

File: web/src/components/storage/ProposalPage.tsx (hunk 3 above)
Comment (imobachgs) [reply web/src/components/storage/ProposalPage.tsx:70]: Yes, I agree. I wanted to get rid of all this stuff but there are already too many changes 😅

File: web/src/components/storage/SpacePolicySelection.tsx (hunk 4 above)
Comment (imobachgs) [reply web/src/components/storage/SpacePolicySelection.tsx:122]: Actually, I think I can remove the `?`.

File: web/src/queries/storage.ts (hunk 6 above)
Comment (imobachgs) [reply web/src/queries/storage.ts:118]: IMHO it belongs here. They are storage-related params that depend on the product.

File: web/src/types/storage.ts (hunk 9 above)
Comment (imobachgs) [reply web/src/types/storage.ts:1]: Yes, we have a plan for unification, but it will take quite some time.

File: web/src/utils.js (hunk 10 above)
Comment (imobachgs) [reply web/src/utils.js:301]: Well, this is a JavaScript file.

File: web/src/components/storage/SpacePolicySelection.tsx (hunk 4 above)
Comment (dgdavid) [reply web/src/components/storage/SpacePolicySelection.tsx]: Yes you can. My brain deleted it automatically when reading it a couple of minutes ago.

File: web/src/utils.js (hunk 10 above)
Comment (dgdavid) [reply web/src/utils.js:301]: Ah, right! I overlooked that tiny detail :P

**Pull request merged**